File: dart/for-each.dart

var sections = [
	{
		'title': 'Getting Started',
		'resetLessonPosition': false,
		'lessons': [
			{'name': 'Welcome'}, 
			{'name': 'Installation'}
		]
	},
	{
		'title': 'Basic Operator',
		'resetLessonPosition': false,
		'lessons': [
			{'name': 'Addition / Substraction'}, 
			{'name': 'Multiplication / Division'}
		]
	}, 
	{
		'title': 'Advanced Topics',
		'resetLessonPosition': true,
		'lessons': [
			{'name': 'Mutability'}, 
			{'name': 'Immutability'}
		]
	}
];

void main() {
	var sectionCounter = 0;
	var lessonCounter  = 0;

	for (var section in sections) {
		if (section['resetLessonPosition'] as bool) {
			lessonCounter = 0;
		}
		section['position'] = ++sectionCounter;

		for (var lesson in section['lessons'] as List) {
			lesson['position'] = (++lessonCounter).toString();
		}
	}
	print(sections);
}

File: javascript/forEach.js
const sections = [
  {
    "title": "Getting started",
    "reset_lesson_position": false,
    "lessons": [
      {"name": "Welcome"},
      {"name": "Installation"}
    ]
  },

  {
    "title": "Basic operator",
    "reset_lesson_position": false,
    "lessons": [
      {"name": "Addition / Subtraction"},
      {"name": "Multiplication / Division"}
    ]
  },

  {
    "title": "Advanced topics",
    "reset_lesson_position": true,
    "lessons": [
      {"name": "Mutability"},
      {"name": "Immutability"}
    ]
  }
];

let lessonPos = 1;
sections.forEach((section, idx) => {
  if (section.reset_lesson_position) lessonPos = 1;
  section.position = idx + 1;
  section.lessons.forEach(lesson => lesson.position = lessonPos++);
});

console.dir(sections, {depth:3});
